http: Allow http filters to add a body that's not readily available to a headers-only request/response

yosrym93 opened the pull request:
Commit Message:
Allow HTTP filters to add a body that's not readily available to a headers-only request/response.
Signed-off-by: Yosry Ahmed <yosryahmed@google.com>

Additional Description:
This is done through adding a new FilterHeadersStatus::ContinueAndDontEndStream. This allows a filter to continue headers iteration without ending the stream. The filter can then add the request/response body by using injectEncodedDataToFilterChain/injectDecodedDataToFilterChain.

Risk Level: ?
Testing: Unit and integration tests.
Docs Changes: N/A
Release Notes: N/A
Fixes #12814

yosrym93 commented on include/envoy/http/filter.h:
```diff
@@ -36,6 +36,15 @@ enum class FilterHeadersStatus {
   // results in creating a header only request/response.
   // This status MUST NOT be returned by decodeHeaders() when end_stream is set to true.
   ContinueAndEndStream,
+  // Continue iteration to remaining filters, but do not end the stream.
+  //
+  // Used when a filter wants to add a body to headers-only request/response, but this body is not
+  // readily available.
+  // This causes the headers iteration to continue, but the stream does not end.
+  // The filter is responsible to continue the stream by providing a body through
+  // injectEncodedDataToFilterChain()/injectDecodedDataToFilterChain().
+  // If the filter cannot provide a body the stream should be reset.
+  ContinueAndDontEndStream,
```
The current implementation supports adding a body only through `inject...`. Adding a body through `addEncodedData`/`addDecodedData` + `continueEncoding`/`continueDecoding` can be supported be adding a new `IterationState`, for e.g. `IterationState::ContinueAndDontEndStream` or `IterationState::WaitingForBody`.

Currently the asserts in `FilterManager::commonContinue()` and `FilterManager::allowIteration()` prevent this, but I tried commenting the asserts and it should work. So all we need to do to support this is to add a new `IterationState`. Do you think we should do it now? or add a TODO to do it if a filter needs it in the future?

Also, can you confirm the correctness of this line ` If the filter cannot provide a body the stream should be reset.`?

yosrym93 replied:
Of course I'll add another unit test for this if you see that we should support it.

mattklein123 replied:
I think just supporting inject for now is OK, but please add a bunch of comments/TODOs and make sure it fails in some easily understood way via ASSERT, etc.

> Also, can you confirm the correctness of this line If the filter cannot provide a body the stream should be reset.?

Yes I think this is correct.

yosrym93 commented:
@mattklein123 This is a draft PR for the feature we discussed in #12814. Please take a look and let me know what you think. I added a unit test, and I'll add integration tests once we agree on the implementation.

yosrym93 commented on source/common/http/filter_manager.cc:
```diff
@@ -123,6 +124,11 @@ bool ActiveStreamFilterBase::commonHandleAfterHeadersCallback(FilterHeadersStatu
     // but continue filter iteration so we actually write the headers/run the cleanup code.
     headers_only = true;
     ENVOY_STREAM_LOG(debug, "converting to headers only", parent_);
+  } else if (status == FilterHeadersStatus::ContinueAndDontEndStream) {
```
This whole else if could be replaced with a `switch`. Is there a particular reason why a `switch` isn't used?

mattklein123 replied:
Probably because the states have been added incrementally over time. Feel free to change to a switch if you like.

yosrym93 commented:
I updated the doc of the new FilterHeadersStatus to explain that using `add...` is not supported yet, with a TODO on how to support it. I also added a message to the ASSERT that fails to make it clear that it fails because continue is called while the `IterationState` is already `Continue`. Hopefully this makes it clear.
Working on the integration test now.

yosrym93 commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -5070,9 +5070,8 @@ TEST_F(HttpConnectionManagerImplTest, FilterContinueAndEndStreamHeaders) {
   conn_manager_->onData(fake_input, true);
 
-  EXPECT_CALL(*encoder_filters_[1], encodeHeaders(_, true))
+  EXPECT_CALL(*encoder_filters_[1], encodeHeaders(_, false))
```
The ASSERTs I added to make sure `ContinueAndEndStream` and `ContinueDontEndStream` are used correctly by filters failed on this test. After reading it, it seems like this was a bug. It makes since that `end_stream` is initially `false`, then the filter changes it to `true` by returning `ContinueAndEndStream`.

The updated version is also consistent with the other versions of the test,  `FilterContinueAndEndStreamData` and `FilterContinueAndEndStreamTrailers`. Which makes me believe more that this was a bug.

mattklein123 replied:
Yes agreed this looks like a bug.

yosrym93 commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -5070,9 +5070,8 @@ TEST_F(HttpConnectionManagerImplTest, FilterContinueAndEndStreamHeaders) {
-  EXPECT_CALL(*encoder_filters_[1], encodeHeaders(_, true))
+  EXPECT_CALL(*encoder_filters_[1], encodeHeaders(_, false))
       .WillOnce(Return(FilterHeadersStatus::ContinueAndEndStream));
-  EXPECT_CALL(*encoder_filters_[1], encodeComplete());
```
After fixing the test I realized that this EXPECT fails. After looking at the code, specifically `FilterManager::encodeHeaders`, it turns out the current behavior is that if a filter returns `ContinueAndEndStream` from `encodeHeaders` (to switch `end_stream` to true), `encodeComplete` is not called for that filter.

This behavior also seems like a bug, but I am not entirely sure. To correct we'll have to edit this condition in `FilterManager::encodeHeaders`:
```
if ((*entry)->end_stream_) {
      (*entry)->handle_->encodeComplete();
}
```
to `if ((*entry)->end_stream_ || state_.encoding_headers_only_)` or update `(*entry)->end_stream_` when a filter changes `end_stream`, either through `ContinueAndEndStream` or `ContinueAndDontEndStream`.

mattklein123 replied:
Yes it's definitely a bug. Nice catch. Can you fix this please? We should definitely be calling encodeComplete() once in all cases when encoding is ended.

yosrym93 replied:
Should I do it in this PR or a separate PR?

mattklein123 commented on source/common/http/filter_manager.cc:
```diff
@@ -108,24 +109,38 @@ bool ActiveStreamFilterBase::commonHandleAfter100ContinueHeadersCallback(
+  case FilterHeadersStatus::Continue:
     headers_continued_ = true;
+    break;
+  default:
+    ASSERT(false, "Unrecognized FilterHeadersStatus");
```
The compiler will not compile unless all states are handled, remove this.

yosrym93 replied:
I only added this as a replacement for the ASSERT in the old else. Maybe it was there in case the status is garbage? Anyway I'll remove it.

mattklein123 commented on test/integration/http2_integration_test.cc:
```diff
@@ -1245,6 +1245,34 @@ TEST_P(Http2IntegrationTest, PauseAndResumeHeadersOnly) {
   ASSERT_TRUE(response->complete());
 }
 
+TEST_P(Http2IntegrationTest, ContinueHeadersOnlyInjectBodyFilter) {
```
Can you put this in protocol_integration_test? It would be good to get coverage of all protocol permutations.

yosrym93 commented:
I moved the test to protocol_integration_tests and removed the default switch statement.

I also fixed the bug, done some minimum refactoring in `FilterManager::decodeHeaders` to be consistent with `FilterManager::encodeHeaders`, and updated the `FilterContinueAndEndStream*` tests in conn_manager_impl_test.

yosrym93 commented:
/retest

mattklein123 reviewed (dismissed):
Thanks looks great. @snowp can you PTAL since you have been doing a bunch of work in this area recently? Thank you.

snowp commented on include/envoy/http/filter.h:
```diff
@@ -36,6 +36,25 @@ enum class FilterHeadersStatus {
+  // Currently, when a filter returns this, the IterationState is Continue. This causes ASSERTs at
+  // FilterManager::commonContinue() to fail when continueDecoding()/continueEncoding() is called;
+  // due to trying to continue iteration when the IterationState is already Continue.
+  ContinueAndDontEndStream,
```
Generally with the docs here I'd say we should make sure that its clear how this differs from Continue (when would you use one over the other) and exactly when it can be used.

yosrym93 replied:
In the docs above it says: 
```
Continue iteration to remaining filters, but do not end the stream.
Used when a filter wants to add a body to headers-only request/response, but this body is not readily available.
```
Do you this is not clear enough? Do you have suggestions in improving the wording here?

snowp replied:
I guess the my reading of it made it sound more of a suggested use of this return value than an ASSERT enforced invariant. Compare the wording with `ContinueAndEndStream` or `StopIteration` the language is a lot stronger, so it might be good to be consistent?

How about something along the lines of

```
Continue iteration to remaining filters, but delay ending the stream. This status MUST NOT be returned when end_stream is already set to false.

This value is intended for filters that want to add a body to a request/response that only contains headers, but this body is not readily available. Delaying end_stream allows the filter to add the body once it's available without blocking.
```

(Feel free to reword, just a suggestion)

snowp commented on source/common/http/filter_manager.cc:
```diff
@@ -902,16 +929,28 @@ void FilterManager::encodeHeaders(ActiveStreamEncoderFilter* filter, ResponseHea
+    const auto continue_iteration = (*entry)->commonHandleAfterHeadersCallback(
+        status, end_stream, state_.encoding_headers_only_);
+
+    // If this filter ended the stream, encodeComplete() should be called for it.
+    if ((*entry)->end_stream_ || state_.encoding_headers_only_) {
+      (*entry)->handle_->encodeComplete();
+    }
```
Is this change related to making the ContinueAndDontEndStream path work? Seems unrelated but maybe I'm missing something

yosrym93 replied:
This is unrelated. It is a small bug I found while working on this (refer to the above conversation with Matt). It is actually related to the `ContinueAndEndStream` path. This makes sure that `encodeComplete()` is called for the filter that returns `ContinueAndEndStream`.

snowp commented on test/integration/filters/continue_headers_only_inject_body_filter.cc:
```diff
@@ -0,0 +1,47 @@
+  }
+
+private:
+  std::string body_ = "body";
```
nit: just make this another constexpr?

snowp commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -5182,6 +5187,63 @@ TEST_F(HttpConnectionManagerImplTest, FilterContinueAndEndStreamTrailers) {
   decoder_filters_[1]->callbacks_->encodeTrailers(std::move(response_trailers));
 }
 
+// filter continues headers iteration without ending the stream, then injects a body later.
```
nit: capitalize `filter`

snowp reviewed (changes requested):
Thanks this seems like a great addition, just a few comments

yosrym93 commented on include/envoy/http/filter.h:
```diff
@@ -36,6 +36,25 @@ enum class FilterHeadersStatus {
   // results in creating a header only request/response.
   // This status MUST NOT be returned by decodeHeaders() when end_stream is set to true.
   ContinueAndEndStream,
+  // Continue iteration to remaining filters, but do not end the stream.
```
As mentioned just at the line below, it should only be used when the request/response is headers-only, i.e `end_stream` is true. Is is also enforced by asserts with a meaningful message.

yosrym93 commented on include/envoy/http/filter.h:
```diff
@@ -36,6 +36,25 @@ enum class FilterHeadersStatus {
+  // Used when a filter wants to add a body to headers-only request/response, but this body is not
+  // readily available.
+  // This causes the headers iteration to continue, but the stream does not end.
+  // The filter is responsible to continue the stream by providing a body through
```
Currently, yes. We can support adding a body by calling `addEncodedBody`\ `addDecodedBody` then calling `continueEncoding`\`continueDecoding`, but Matt said there's no need to implement it for now. I also explained how it could be implemented in the TODO below.

I think calling `continueEncoding`\`continueDecoding` without adding a body should be an incorrect behavior as downstream filters will receive headers with `end_stream = false` then no body. This could cause some filters to do an incorrect behavior as they would assume there's an incoming body.  Right now I documented that the filter should reset the stream if it was unable to provide a body.

snowp replied:
Is it possible to add an ASSERT that prevents continue* from being called in this case? Or is there already one that will trigger?

Like you said I don't think there's a need to implement this but it would be great to have the current state of what is possible/not possible documented + ASSERTed on.

I think is another wording thing like my other comment: it would be good to explicitly call out that calling continue* won't work.

Thinking about this some more it might also be nice to point out that the filter probably doesn't have to explicitly reset, it can just rely on idle timeouts etc. to reset it.

yosrym93 replied:
There's already an ASSERT which I added a message to make it a little bit clearer. I will update the documentation to make things more explicit (considering the other comment too). As for the reset part, I guess it's more future-proof and clearer to have the filter explicitly reset the stream, as in: "We failed to provide the body, this response failed, reset the stream" rather than "We failed to provide a body, do nothing".

yosrym93 replied:
I updated the comments to be more explicit.

snowp commented on include/envoy/http/filter.h:
```diff
@@ -32,28 +32,42 @@ enum class FilterHeadersStatus {
   StopIteration,
-  // Continue iteration to remaining filters, but ignore any subsequent data or trailers. This
+  // Continue headers iteration to remaining filters, but ignore any subsequent data or trailers.
+  // This
```
reflow line here

snowp commented on include/envoy/http/filter.h:
```diff
@@ -32,28 +32,42 @@ enum class FilterHeadersStatus {
+  // watermark events when injecting a body, see:
+  // https://github.com/envoyproxy/envoy/blob/master/source/docs/flow_control.md.
+  //
+  // The last call to inject a body MUST have end_stream set to true to conclude the stream.
```
Presumably it could also be ended with `addEncoded/DecodedTrailers`?

yosrym93 replied:
Right. I will add that.

yosrym93 replied:
Actually I looked at `addEncoded/DecodedTrailers` docs and code. it looks like they can only be called in `encodeData` when `end_stream` is true. So they don't apply here. Will revert the last change.

yosrym93 replied:
The current API does not support adding trailers in this scenario, it would need similar `injectTrailers*` methods.

yosrym93 replied:
I documented this properly.

snowp replied:
Thanks for looking into it!

snowp reviewed (changes requested):
Just a few more comments, thanks for iterating!

snowp reviewed (approved):
LGTM, thanks!

mattklein123 commented:
Awesome. Can you merge main? It was broken temporarily.

/wait

yosrym93 commented:
> Awesome. Can you merge main? It was broken temporarily.
> 
> /wait

Done!
